wallet: Use correct effective value when checking target

aureleoules opened the pull request:
Fixes #26185. The following assert failed because it was not checked in the parent function.

https://github.com/bitcoin/bitcoin/blob/2bd9aa5a44b88c866c4d98f8a7bf7154049cba31/src/wallet/coinselection.cpp#L391

aureleoules committed:
wallet: Use correct effective value when checking target

aureleoules commented:
@1440000bytes could you check if this fixes your issue? Thanks.

fanquake commented:
> Also includes a quick refactor.

Generally, we avoid throwing "quick refactors" into bug fixes. Especially if they are going to be backported. Or, at least the refactor should be a separate commit, so if it doesn't need backporting, it doesn't have to be.

aureleoules committed:
test: Check external coin effective value is used in CoinSelection

aureleoules commented:
> Generally, we avoid throwing "quick refactors" into bug fixes.

aureleoules commented:
Added a unit test.

1440000bytes commented:
@aureleoules  I want to clarify one:

Is this a bug only on Windows or other platforms? If its a bug on some linux distro as well it will be easier for me to test

1440000bytes commented:
> @aureleoules I want to clarify one thing:
> 
> Is this a bug only on Windows or other platforms? If its a bug on some linux distro as well it will be easier for me to test

Or maybe that should be job of testers.. Give me one day I will test it.

1440000bytes commented:
I don't understand whats happening in this issue. I compiled PR branch and tested it on Windows, it fixes the previous error and bitcoind or bitcoin-qt does not crash. However, returns "insufficient funds".

@sipa said this is platform independent but I tried to run the same command on Fedora with v24.0rc1 and it returns PSBT with a change:

Windows (PR Branch):

```
bitcoin-cli listunspent

[
  {
    "txid": "e911734bf711009dd6057c88bc598df2d31c18d1cc6bd6c790a704b519d77e68",
    "vout": 0,
    "address": "2N7uBTqqQ99ArbgmBuw2AANsf24bMVdYxgp",
    "label": "",
    "redeemScript": "522102de62833c851509cf13c0c63e0dc3ef7b74f7d86b17e9090fab68c09830cb2f2e2103cf11a2bc1371dad3691bce59e37d85b12aae6b030b533d4e7c5aedeb47a5ec0e52ae",
    "scriptPubKey": "a914a0bf75baf8f3a4ae761c2faa2516b8540e993d8987",
    "amount": 0.00100000,
    "confirmations": 595,
    "spendable": false,
    "solvable": true,
    "desc": "sh(multi(2,[aa130fed]02de62833c851509cf13c0c63e0dc3ef7b74f7d86b17e9090fab68c09830cb2f2e,[00890d8d]03cf11a2bc1371dad3691bce59e37d85b12aae6b030b533d4e7c5aedeb47a5ec0e))#j8w6g6zg",
    "parent_descs": [
    ],
    "safe": true
  }
]

bitcoin-cli walletcreatefundedpsbt "[{\"txid\":\"e911734bf711009dd6057c88bc598df2d31c18d1cc6bd6c790a704b519d77e68\",\"vout\":0}]" "[{\"tb1q9fxp2p0sm93jdlul8v3svzn7x69zcd9ljsgans\":\"0.0009\"}]"

Insufficient funds (code -4)
```

Fedora (v24.0rc1):

```
bitcoin-cli listunspent

[
  {
    "txid": "e911734bf711009dd6057c88bc598df2d31c18d1cc6bd6c790a704b519d77e68",
    "vout": 0,
    "address": "2N7uBTqqQ99ArbgmBuw2AANsf24bMVdYxgp",
    "label": "",
    "redeemScript": "522102de62833c851509cf13c0c63e0dc3ef7b74f7d86b17e9090fab68c09830cb2f2e2103cf11a2bc1371dad3691bce59e37d85b12aae6b030b533d4e7c5aedeb47a5ec0e52ae",
    "scriptPubKey": "a914a0bf75baf8f3a4ae761c2faa2516b8540e993d8987",
    "amount": 0.00100000,
    "confirmations": 595,
    "spendable": false,
    "solvable": true,
    "desc": "sh(multi(2,[aa130fed]02de62833c851509cf13c0c63e0dc3ef7b74f7d86b17e9090fab68c09830cb2f2e,[00890d8d]03cf11a2bc1371dad3691bce59e37d85b12aae6b030b533d4e7c5aedeb47a5ec0e))#j8w6g6zg",
    "parent_descs": [
    ],
    "safe": true
  }
]

bitcoin-cli  walletcreatefundedpsbt "[{\"txid\":\"e911734bf711009dd6057c88bc598df2d31c18d1cc6bd6c790a704b519d77e68\",\"vout\":0}]" "[{\"tb1q9fxp2p0sm93jdlul8v3svzn7x69zcd9ljsgans\":\"0.0009\"}]"

{
  "psbt": "cHNidP8BAHECAAAAAWh+1xm1BKeQx9ZrzNEYHNPyjVm8iHwF1p0AEfdLcxHpAAAAAAD9////AsYlAAAAAAAAFgAUJT0nNy2sgypq59m0DmCbgIWp0m6QXwEAAAAAABYAFCpMFQXw2WMm/587IwYKfjaKLDS/AAAAAAABAHICAAAAAQzC3aNgVZV/fAJNlVsvxAMFuTtvRKqlPocBAGlOSHxDAAAAAAD+////AqCGAQAAAAAAF6kUoL91uvjzpK52HC+qJRa4VA6ZPYmHm52Yh1IGAAAWABQN4fccLGiwTrWIb1+7fsWapgno0aOsAQABBEdSIQLeYoM8hRUJzxPAxj4Nw+97dPfYaxfpCQ+raMCYMMsvLiEDzxGivBNx2tNpG85Z432FsSquawMLUz1OfFrt60el7A5SriIGAt5igzyFFQnPE8DGPg3D73t099hrF+kJD6towJgwyy8uBKoTD+0iBgPPEaK8E3Ha02kbzlnjfYWxKq5rAwtTPU58Wu3rR6XsDgQAiQ2NACICAiOsbsXVfUGqw6ybbE+LIjii6xEJVH7K9GgwIxu6pruVEMw4p/EAAACAAQAAgAcAAIAAAA==",
  "fee": 0.00000330,
  "changepos": 0
}

```

glozow commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -922,5 +922,52 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
 
+BOOST_AUTO_TEST_CASE(SelectCoins_effective_value_test)
+{
+    // Test that effective value is used in coin selection
+    // This test creates an external coin with a high effective value
+    // The coin will not be selected because its effective value is less than the target
```
No coin selection is actually happening, as `m_allow_other_inputs` = false. Being high can sometimes be a problem, but in this case the effective value is too *low*.

```suggestion
    // Test that effective value is used to check the value of preset inputs.
    // This test creates a coin in which the effective value is too low to cover the fees, but is higher than the target value.
    // The coin is selected using coin control, with m_allow_other_inputs = false. SelectCoins should fail.
```

Xekyo replied:
The effective value of a UTXO is `effValue = value - fee`, so rather than the effective value being too low to cover the fees, the effective value is too low to create the output after deducting the fees:

```suggestion
    // Test that the effective value is used to check whether preset inputs provide sufficient funds when subtract_fee_outputs is not used.
    // This test creates a coin whose value is higher than the target but whose effective value is lower than the target.
    // The coin is selected using coin control, with m_allow_other_inputs = false. SelectCoins should fail due to insufficient funds.
```

glozow commented:
@1440000bytes  So IIUC you have 1 coin worth 0.001 BTC (as shown by your `listunspent`). You tried to create a transaction paying 0.0009 using `walletcreatefundedpsbt`, passing that coin in manually. Since you specified an input manually, `m_allow_other_inputs` is false.
https://github.com/bitcoin/bitcoin/blob/f59e91511a3aa8b2770eeec7034ddc1a9dec918b/src/wallet/rpc/spend.cpp#L1659-L1663
This needs to fund the payment and the fees for the transaction, including the cost to spend the coin. 0.001 covers the payment and non-input fees, but not the cost to spend. So it is indeed the case that the wallet has insufficient funds.

It seems like the cause of the crash was:

`SelectionResult::AddInput` changes `m_use_effective` to true:
https://github.com/bitcoin/bitcoin/blob/f59e91511a3aa8b2770eeec7034ddc1a9dec918b/src/wallet/spend.cpp#L584
https://github.com/bitcoin/bitcoin/blob/f59e91511a3aa8b2770eeec7034ddc1a9dec918b/src/wallet/coinselection.cpp#L441-L445
`GetSelectedValue()` is 0.001 and `GetSelectedEffectiveValue()` is something less than 0.0009.
So this condition doesn't hit:
https://github.com/bitcoin/bitcoin/blob/f59e91511a3aa8b2770eeec7034ddc1a9dec918b/src/wallet/spend.cpp#L585
But the assertion does, since `selected_effective_value` is something less than 0.0009.

And thus why adding the `m_use_effective` case here works.

If you're getting different results... are you sure it isn't any difference in the config? If the Fedora one paid 0.00000330 in fees, I guess this assertion wouldn't hit. It seems plausible that the fees would be automatically set higher than that.

glozow reviewed (commented):
concept ACK 1e60f5d4d5e9990e00ddbaa0cfacf3c5731d73f9
whether or not it's the cause of the crash in #26185, this is a bug fix, as AFAICT it's incorrect to not use effective value here

Xekyo commented:
> I don't understand whats happening in this issue. I compiled PR branch and tested it on Windows, it fixes the previous error and bitcoind or bitcoin-qt does not crash. However, returns "insufficient funds".
> 
> @sipa said this is platform independent but I tried to run the same command on Fedora with v24.0rc1 and it returns PSBT with a change:

@1440000bytes: I missed this at first, but could you confirm that the first output you posted is resulting from running `PR/26203` (this branch) on Windows, and the second output is from running `v24.0rc1` on Fedora?

mzumsande commented:
Concept ACK.

The assert can also be reproduced by adjusting the amount in the psbt just between the regimes where it succeeds and the regime with insufficient funds. E.g. on an empty regtest datadir:
```
./bitcoind -regtest -fallbackfee=0.001
./bitcoin-cli -regtest -generate 101
./bitcoin-cli -regtest listunspent -> use txid in next call
./bitcoin-cli -regtest walletcreatefundedpsbt "[{\"txid\":\"<txid>\",\"vout\":0}]" "[{\"<some address>\":\"49.9999\"}]"
```
will lead to an assert on master (which is fixed by this PR).

1440000bytes commented:
> This needs to fund the payment and the fees for the transaction, including the cost to spend the coin. 0.001 covers the payment and non-input fees, but not the cost to spend. So it is indeed the case that the wallet has insufficient funds.

What is cost to spend?

> @1440000bytes: I missed this at first, but could you confirm that the first output you posted is resulting from running PR/26203 (this branch) on Windows, and the second output is from running v24.0rc1 on Fedora?

Yes

and both have same config:

```
# Windows

signet=1
txindex=1

signet.rpcport=18332
rpcuser=user
rpcpassword=pass

fallbackfee=0.0004


# Fedora

signet=1
txindex=1

[signet] 
rpcport=18332
rpcuser=user
rpcpassword=pass

fallbackfee=0.0004

```

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -582,7 +582,13 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& a
         result.AddInput(preset_inputs);
-        if (result.GetSelectedValue() < nTargetValue) return std::nullopt;
+
+        if (result.UseEffectiveValue() && result.GetSelectedEffectiveValue() < nTargetValue) {
+            return std::nullopt;
+        } else if (result.GetSelectedValue() < nTargetValue) {
+            return std::nullopt;
+        }
```
We always use the effective value of UTXOs except when the recipient pays the fees via SFFO, so you can go by the negation of the corresponding coin selection parameter:

```suggestion
        if (!coin_selection_params.m_subtract_fee_outputs && result.GetSelectedEffectiveValue() < nTargetValue) {
            return std::nullopt;
        } else if (result.GetSelectedValue() < nTargetValue) {
            return std::nullopt;
        }
```

Perhaps it would be nicer if we dropped the negation and swapped `if` and `else if` branches:

```suggestion
        if (coin_selection_params.m_subtract_fee_outputs && result.GetSelectedValue() < nTargetValue) {
            return std::nullopt;
        } else if (result.GetSelectedEffectiveValue() < nTargetValue) {
            return std::nullopt;
        }
```

or more compactly:

```suggestion
        if (result.GetSelectedEffectiveValue() < nTargetValue || coin_selection_params.m_subtract_fee_outputs && result.GetSelectedValue() < nTargetValue) {
            return std::nullopt;
        }
```

aureleoules replied:
Thanks, I pushed your first edited version for readability, the two others translate to another statement that is invalid though.

murchandamus reviewed (commented):
Concept ACK, have a couple small improvement suggestions:

1440000bytes commented:
@MarcoFalke 

Use marcofalke... Don't do anything else I will kil you...

aureleoules commented:
Thank you @Xekyo @glozow @mzumsande for the reviews, I addressed your comments. Thanks @1440000bytes for testing.
push diff here: https://github.com/bitcoin/bitcoin/compare/1e60f5d4d..d0d9cf7aea2ff36a14a19e6999400e4070b7b0c9

1440000bytes commented:
> Thank you @Xekyo @glozow @mzumsande for the reviews, I addressed your comments. Thanks @1440000bytes for testing. push diff here: https://github.com/bitcoin/bitcoin/compare/1e60f5d4d..d0d9cf7aea2ff36a14a19e6999400e4070b7b0

People will respond. I am wrong but but will do.

S3RK commented:
Overall the change looks correct to me. I was trying to backport the test to see where the bug was introduced, but didn't succeed due to too many refactors. Is it a regression from 23.0 or not?

glozow reviewed (commented):
reACK d0d9cf7aea2ff36a14a19e6999400e4070b7b0c9

furszy reviewed (approved):
ACK d0d9cf7a

Happily, #25685 fixes this too. I introduced the same preset-inputs result selection target check inside aa5fdbad for another reason there. So.. the unit test introduced here is also passing there without having to cherry-pick the bug fix 76b79c1a.

glozow commented:
> Happily, https://github.com/bitcoin/bitcoin/pull/25685 fixes this too.

That's good to hear, though I think this is more appropriate for a backport.
#26242

murchandamus reviewed (commented):
Post-merge ACK d0d9cf7aea2ff36a14a19e6999400e4070b7b0c9
